src/mesh/mesh_modification.C

```diff
@@ -1834,29 +1801,18 @@ void MeshTools::Modification::flatten(MeshBase & mesh)
     mesh.delete_elem(elem);
 
   // Add the copied (now level-0) elements back to the mesh
-  {
-    for (std::vector<Elem *>::iterator it = new_elements.begin();
-         it != new_elements.end();
-         ++it)
-      {
-#ifndef NDEBUG
-        dof_id_type orig_id = (*it)->id();
-
-        // ugly mid-statement endif to avoid unused variable warnings
-        Elem * added_elem =
-#endif
-          mesh.add_elem(*it);
+  for (auto & new_elem : new_elements)
+    {
+      Elem * added_elem = mesh.add_elem(new_elem);
 
-#ifndef NDEBUG
-        dof_id_type added_id = added_elem->id();
-#endif
+      // If the Elem, as it was re-added to the mesh, now has a
+      // different ID (this is unlikely, so it's just an assert)
+      // the boundary information will no longer be correct.
+      libmesh_assert_equal_to (new_elem->id(), added_elem->id());
```

Comment:
This actually isn't equivalent to the original uglier code.  IIRC we had a bug at one point where adding a new element changed the id of that element.  We have to save the original id before add_elem to be sure that hasn't happened.

Reply:
OK, I can fix that.